packages/client/src/publish/Publisher.ts

```diff
@@ -1,65 +1,171 @@
+        this.cacheConfig = cacheConfig
+        this.getMessageFactory = pLimitFn(pMemoize(async (streamId: StreamID) => {  // TODO is it better to use pMemoize or CacheAsyncFn (e.g. after revoked publish permissions?)
+            return this.createMessageFactory(streamId)
+        }, {
+            cacheKey: ([streamId]) => streamId
+        }))
         this.getGroupKeyQueue = pMemoize(async (streamId: StreamID) => {
             return new GroupKeyQueue(streamId, groupKeyStore)
         }, {
             cacheKey: ([streamId]) => streamId
         })
     }
 
-    async publish<T>(streamDefinition: StreamDefinition, content: T, metadata?: MessageMetadata): Promise<StreamMessage<T>> {
-        return this.pipeline.publish({
-            streamDefinition,
-            content,
-            timestamp: parseTimestamp(metadata),
-            partitionKey: metadata?.partitionKey,
-            msgChainId: metadata?.msgChainId,
-            messageType: metadata?.messageType,
-            encryptionType: metadata?.encryptionType
+    private async createMessageFactory(streamId: StreamID): Promise<MessageFactory> {
+        const [ stream, authenticatedUser ] = await Promise.all([
+            this.streamRegistryCached.getStream(streamId),
+            this.authentication.getAddress()
+        ])
+        const isPublisher = await this.streamRegistryCached.isStreamPublisher(streamId, authenticatedUser)
+        if (!isPublisher) {
+            throw new Error(`${authenticatedUser} is not a publisher on stream ${streamId}`)
+        }
+        const isPublicStream = await this.streamRegistryCached.isPublic(streamId)
+        const queue = await this.getGroupKeyQueue(streamId)
+        return new MessageFactory({
+            streamId,
+            partitionCount: stream.partitions,
+            isPublicStream,
+            publisherId: authenticatedUser.toLowerCase(),
+            createSignature: (payload: string) => this.authentication.createMessagePayloadSignature(payload),
+            useGroupKey: () => queue.useGroupKey(),
+            cacheConfig: this.cacheConfig
         })
     }
 
-    async start(): Promise<void> {
-        this.pipeline.start()
-    }
-
-    async stop(): Promise<void> {
-        return this.pipeline.stop()
+    async publish<T>(
+        streamDefinition: StreamDefinition,
+        content: T,
+        metadata?: MessageMetadata
+    ): Promise<StreamMessage<T>> {
+        const timestamp = parseTimestamp(metadata)
+        /*
+         * There are some steps in the publish process which need to be done sequentially:
+         * - message chaining
+         * - consuming a group key from a queue
+         * 
+         * It is also good if messages are published to node in the same sequence, as that can avoid
+         * unnecessary gap fills (if a subscriber would receive messages m1, m2, m3 in order m1, m3, m2
+         * it would try to get m2 via a gap fill resend before it receives it normally).
+         * 
```

Comment:
```suggestion
         * It is also good if messages are published to node in the same sequence (within a message chain), as that can avoid
         * unnecessary gap fills (if a subscriber would receive messages m1, m2, m3 in order m1, m3, m2
         * it would try to get m2 via a gap fill resend before it receives it normally).
         * 
```

Reply:
Fixed